Composer/packages/extensions/obiformeditor/src/FormEditor.tsx

```diff
@@ -30,15 +30,16 @@ export interface FormEditorProps extends ShellData {
 }
 
 export const FormEditor: React.FunctionComponent<FormEditorProps> = props => {
-  const { data, schemas, memory, dialogs, shellApi } = props;
+  const { data, schemas, memory, dialogs, shellApi, externalUpdate } = props;
   const [localData, setLocalData] = useState(data);
   const type = getType(localData);
 
   useEffect(() => {
-    if (!isEqual(localData, data)) {
+    // undo and redo will update the localData
+    if (externalUpdate?.scope === UpdateScope.DialogFile && !isEqual(localData, data)) {
       setLocalData(data);
     }
-  }, [data]);
+  }, [data, externalUpdate]);
```

Comment:
I'm confused why the `!isEqual(localData, data)` wasn't sufficient. In general, I'm not a big fan of this `externalUpdate` api. My view is that `localData` is an internal (FormEditor) performance optimization and should be made to work when the global data is updated outside of the form. @lei9444 can you revisit this and see if we can make it work without the `externalUpdate`?

Reply:
Hi, @a-b-r-o-w-n @boydc2014 The purpose I want to add external status is that if the update is from the editor and we update the local value will cause a bug like this:
![test1](https://user-images.githubusercontent.com/39758135/72324867-dc064d80-36e6-11ea-93d4-448425a25196.gif)

So I add a status to tell the extensions there is an update from the shell.

Do you have any better ways to solve the problem?

Reply:
@a-b-r-o-w-n my understanding is that:
1. using a local data is essentially making the data flow single-direction, very similar to how client save data to server today.   And this "externalUpdate" is a little bit like servery says "your data is not the latest, you need to refresh". 
2. the !Equals is not good enough because this !Equals check is actually will always be true when the local data is changing very quickly, like the "typing" scenario @lei9444  metioned above, which caused the "eating byte" bug, right？
3. The `externalUpdate` check would be good enough, i would expect this would replace this "!Equals" check.

is this make sense?

Reply:
I prefer force remount formEditor from ExtensionContainer.tsx when we want some force flush. 
keep the concept of  `externalUpdate` in shell, then formEditor do not have to aware any `externalUpdate` and can keep it's internal value as well.

Reply:
Sorry I was out sick yesterday so haven't had a chance to respond. 

I understand the race-condition that using a simple `isEqual` causes, but since the shell save is debounced, we should have enough time between a keypress and the store being updated for it not to matter.

The option @zhixzhan mentioned is interesting. My only concern would be the need to maintain any local extension state when a flush event occurs.

Reply:
yes, @a-b-r-o-w-n if the store updated quick enough before another key stoke, it would work, and that's the case for today. the only thing missing is, we need to notify the form (in Shell side) be be refreshed, then ExtensionContainer layer will remount form (like @zhixzhan). actually that's the first iteration of Leilei's solution, and in that solution, leilei registered a new api instead of create and keep a new state.

The pros of that one are Form doesn't requires any changes, and all the works in done by Shell.  The cons  are that overall a little more effort and change, compared to the current solution that directly let Form to refresh itself based on this shell state. 

My previous standpoint was that since anyhow this kind of signal that "data is updated elsewhere" is must be aware of the extension (ExtensionContainer+Form), i prefer a simple change that let Form react to that signal.    Based on @a-b-r-o-w-n  and @zhixzhan, i think prefer the changes been made in Shell and ExtensionContainer level, is also a very conceptually clean approach that keep things decoupled.

@lei9444 can you switch to the previous approach, see that more resonates with folks?

Reply:
@a-b-r-o-w-n @boydc2014 @zhixzhan I have added a key to force update the extension if the update is from the shell. Maybe we can add the status to the props in the future, and editors can get the status and judge for themself.